src/test/java/com/google/devtools/build/lib/bazel/rules/genrule/GenRuleWindowsTest.java

```diff
@@ -0,0 +1,142 @@
+
+  private static final Pattern POWERSHELL_COMMAND_PATTERN =
+      Pattern.compile(".*'utf8';\\s+(?<command>.*)");
+
+  private void assertCommandEquals(String expected, String command, Pattern pattern) {
+    // Ensure the command after the genrule setup is correct.
+    Matcher m = pattern.matcher(command);
+    if (m.matches()) {
+      command = m.group("command");
+    }
+    assertThat(command).isEqualTo(expected);
+  }
+
+  private void assertBatchCommandEquals(String expected, String command) {
+    assertCommandEquals(expected, command, BATCH_COMMAND_PATTERN);
+  }
+
+  private void assertPowershellCommandEquals(String expected, String command) {
+    assertCommandEquals(expected, command, POWERSHELL_COMMAND_PATTERN);
+  }
+
+  private static String getWindowsPath(Artifact artifact) {
+    return artifact.getExecPathString().replace('/', '\\');
+  }
+
+  @Test
+  public void testCmdBatchIsPreferred() throws Exception {
+    scratch.file(
+        "genrule1/BUILD",
+        "genrule(name = 'hello_world',",
+        "outs = ['message.txt'],",
+        "cmd  = 'echo \"Hello, default cmd.\" >$(location message.txt)',",
+        "cmd_bash  = 'echo \"Hello, Bash cmd.\" >$(location message.txt)',",
+        "cmd_bat  = 'echo \"Hello, Batch cmd.\" >$(location message.txt)')");
+
+    Artifact messageArtifact = getFileConfiguredTarget("//genrule1:message.txt").getArtifact();
+    SpawnAction shellAction = (SpawnAction) getGeneratingAction(messageArtifact);
+
+    assertThat(shellAction).isNotNull();
+    assertThat(shellAction.getOutputs()).containsExactly(messageArtifact);
+
+    String expected = "echo \"Hello, Batch cmd.\" >" + getWindowsPath(messageArtifact);
+    assertThat(shellAction.getArguments().get(0)).isEqualTo("cmd.exe");
+    assertThat(shellAction.getArguments().get(1)).isEqualTo("/c");
+    assertBatchCommandEquals(expected, shellAction.getArguments().get(2));
+  }
+
+  @Test
+  public void testCmdPsIsPreferred() throws Exception {
+    scratch.file(
+        "genrule1/BUILD",
+        "genrule(name = 'hello_world',",
+        "outs = ['message.txt'],",
+        "cmd  = 'echo \"Hello, default cmd.\" >$(location message.txt)',",
+        "cmd_bash  = 'echo \"Hello, Bash cmd.\" >$(location message.txt)',",
+        "cmd_bat  = 'echo \"Hello, Batch cmd.\" >$(location message.txt)',",
+        "cmd_ps  = 'echo \"Hello, Powershell cmd.\" >$(location message.txt)')");
+
+    Artifact messageArtifact = getFileConfiguredTarget("//genrule1:message.txt").getArtifact();
+    SpawnAction shellAction = (SpawnAction) getGeneratingAction(messageArtifact);
+
+    assertThat(shellAction).isNotNull();
+    assertThat(shellAction.getOutputs()).containsExactly(messageArtifact);
+
+    String expected = "echo \"Hello, Powershell cmd.\" >" + getWindowsPath(messageArtifact);
+    assertThat(shellAction.getArguments().get(0)).isEqualTo("powershell.exe");
+    assertThat(shellAction.getArguments().get(1)).isEqualTo("/c");
```

Comment:
Could you verify that the generated file is indeed utf-8?

Reply:
I don't think the action is executed here. So I cannot check. But in the python test, if I don't set `$PSDefaultParameterValues['*:Encoding'] = 'utf8'`, the generated file will be in utf-16, which caused the test to fail.

Reply:
Can we test in Python that the file is indeed utf-8 and not utf-16?

I mean, if we only test that we pass the right flag to PS it's only as good as testing that you sent a message to someone -- it doesn't verify the recipient actually got your message.

Reply:
I think @ulfjack found such a bug in Bazel a few weeks ago : the integration tests only asserted that Bazel executed the right command, they didn't assert the output of that command.

Reply:
I agreed, I think we already have such a test in genrule_test.py, that is how I discovered we need to set `$PSDefaultParameterValues['*:Encoding'] = 'utf8'`